arithmetization/src/main/java/net/consensys/linea/zktracer/container/stacked/StackedSet.java

```diff
@@ -0,0 +1,134 @@
+  @Getter private final Set<E> operationSinceBeginningOfTheConflation;
+  @Getter private final Set<E> thisTransactionOperation;
+  private final CountOnlyOperation lineCounter = new CountOnlyOperation();
+  @Getter private boolean conflationFinished = false;
+
+  public StackedSet() {
+    operationSinceBeginningOfTheConflation = new HashSet<>();
+    thisTransactionOperation = new HashSet<>();
+  }
+
+  /** Prefer this constructor as we preallocate more needed memory */
+  public StackedSet(
+      final int expectedConflationNumberOperations, final int expectedTransactionNumberOperations) {
+    operationSinceBeginningOfTheConflation = new HashSet<>(expectedConflationNumberOperations);
+    thisTransactionOperation = new HashSet<>(expectedTransactionNumberOperations);
+  }
+
+  /**
+   * When we enter transaction n, the previous transaction n-1 {@link thisTransactionOperation} (or
+   * an empty set if n == 0) is definitely added to the set of all transactions {@link
+   * operationSinceBeginningOfTheConflation}. We reset {@link thisTransactionOperation} to an empty
+   * set as it now represents the operations added at transaction n.
+   */
+  public void enter() {
+    operationSinceBeginningOfTheConflation.addAll(thisTransactionOperation);
+    thisTransactionOperation.clear();
+    lineCounter.enter();
+  }
+
+  public void pop() {
+    thisTransactionOperation.clear();
+    lineCounter.pop();
+  }
+
+  public int size() {
+    return thisTransactionOperation.size() + operationSinceBeginningOfTheConflation.size();
+  }
+
+  public int lineCount() {
+    return lineCounter.lineCount();
+  }
+
+  public Set<E> getAll() {
+    Preconditions.checkState(conflationFinished, "Conflation not finished");
+    return operationSinceBeginningOfTheConflation;
+  }
+
+  public boolean isEmpty() {
+    return size() == 0;
+  }
+
+  public boolean contains(Object o) {
+    return thisTransactionOperation.contains(o)
+        || operationSinceBeginningOfTheConflation.contains(o);
+  }
+
+  public boolean add(E e) {
+    if (!operationSinceBeginningOfTheConflation.contains(e)) {
+      final boolean isNew = thisTransactionOperation.add(e);
+      if (isNew) {
+        lineCounter.add(e.lineCount());
+      }
+      return isNew;
+    }
+    return false;
+  }
+
+  public boolean containsAll(@NotNull Collection<?> c) {
+    for (var x : c) {
+      if (!contains(x)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public boolean addAll(@NotNull Collection<? extends E> c) {
+    boolean r = false;
+    for (var x : c) {
+      r |= add(x);
+    }
+    return r;
+  }
+
+  public void clear() {
+    operationSinceBeginningOfTheConflation.clear();
+    thisTransactionOperation.clear();
+    lineCounter.clear();
+  }
+
+  public void finishConflation() {
+    conflationFinished = true;
+    operationSinceBeginningOfTheConflation.addAll(thisTransactionOperation);
+    thisTransactionOperation.clear();
```

Comment:
What about the lineCounter here ? shall we call enter on lineCounter to be sure we add the line counting of the last transaction (in the block) ?

Reply:
we could add it for clarity, but there is no need, as lineCount always return the sum; I can do it for clarity if you prefer.

Reply:
There is no need in this case. May be just add a comment for the class, regarding enter and finishConflation because it is not clear without context (callers context).

Reply:
done